--- src/core/lib/iomgr/tcp_posix.cc
+++ src/core/lib/iomgr/tcp_posix.cc
@@ -415,7 +415,6 @@ static void call_read_cb(grpc_tcp* tcp, grpc_error* error) {
 #define MAX_READ_IOVEC 4
 static void tcp_do_read(grpc_tcp* tcp) {
   GPR_TIMER_SCOPE("tcp_do_read", 0);
-  struct msghdr msg;
   struct iovec iov[MAX_READ_IOVEC];
   ssize_t read_bytes;
   size_t i;
@@ -427,13 +426,9 @@ static void tcp_do_read(grpc_tcp* tcp) {
     iov[i].iov_len = GRPC_SLICE_LENGTH(tcp->incoming_buffer->slices[i]);
   }
 
-  msg.msg_name = nullptr;
-  msg.msg_namelen = 0;
+  struct msghdr msg = {};
   msg.msg_iov = iov;
   msg.msg_iovlen = static_cast<msg_iovlen_type>(tcp->incoming_buffer->count);
-  msg.msg_control = nullptr;
-  msg.msg_controllen = 0;
-  msg.msg_flags = 0;
 
   GRPC_STATS_INC_TCP_READ_OFFER(tcp->incoming_buffer->length);
   GRPC_STATS_INC_TCP_READ_OFFER_IOV_SIZE(tcp->incoming_buffer->count);
@@ -682,12 +677,8 @@ static bool process_errors(grpc_tcp* tcp) {
     struct iovec iov;
     iov.iov_base = nullptr;
     iov.iov_len = 0;
-    struct msghdr msg;
-    msg.msg_name = nullptr;
-    msg.msg_namelen = 0;
+    struct msghdr msg = {};
     msg.msg_iov = &iov;
-    msg.msg_iovlen = 0;
-    msg.msg_flags = 0;
 
     union {
       char rbuf[1024 /*CMSG_SPACE(sizeof(scm_timestamping)) +
@@ -791,7 +782,6 @@ static void tcp_handle_error(void* arg /* grpc_tcp */, grpc_error* error) {
 #define MAX_WRITE_IOVEC 1000
 #endif
 static bool tcp_flush(grpc_tcp* tcp, grpc_error** error) {
-  struct msghdr msg;
   struct iovec iov[MAX_WRITE_IOVEC];
   msg_iovlen_type iov_size;
   ssize_t sent_length;
@@ -824,19 +814,14 @@ static bool tcp_flush(grpc_tcp* tcp, grpc_error** error) {
     }
     GPR_ASSERT(iov_size > 0);
 
-    msg.msg_name = nullptr;
-    msg.msg_namelen = 0;
+    struct msghdr msg = {};
     msg.msg_iov = iov;
     msg.msg_iovlen = iov_size;
-    msg.msg_flags = 0;
     if (tcp->outgoing_buffer_arg != nullptr) {
       if (!tcp_write_with_timestamps(tcp, &msg, sending_length, &sent_length,
                                      error))
         return true; /* something went wrong with timestamps */
     } else {
-      msg.msg_control = nullptr;
-      msg.msg_controllen = 0;
-
       GRPC_STATS_INC_TCP_WRITE_SIZE(sending_length);
       GRPC_STATS_INC_TCP_WRITE_IOV_SIZE(iov_size);
 
