--- src/core/lib/iomgr/socket_utils_common_posix.cc
+++ src/core/lib/iomgr/socket_utils_common_posix.cc
@@ -30,7 +30,9 @@
 #include <fcntl.h>
 #include <limits.h>
 #include <netinet/in.h>
+#ifndef GPR_CLOUDABI
 #include <netinet/tcp.h>
+#endif
 #include <stdio.h>
 #include <string.h>
 #include <sys/socket.h>
--- src/core/lib/iomgr/tcp_server_posix.cc
+++ src/core/lib/iomgr/tcp_server_posix.cc
@@ -32,7 +32,9 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <netinet/in.h>
+#ifndef GPR_CLOUDABI
 #include <netinet/tcp.h>
+#endif
 #include <string.h>
 #include <sys/socket.h>
 #include <sys/stat.h>
--- src/core/lib/iomgr/udp_server.cc
+++ src/core/lib/iomgr/udp_server.cc
@@ -37,7 +37,9 @@
 #include <fcntl.h>
 #include <limits.h>
 #include <netinet/in.h>
+#ifndef GPR_CLOUDABI
 #include <netinet/tcp.h>
+#endif
 #include <string.h>
 #include <sys/socket.h>
 #include <sys/stat.h>
