--- ares_getopt.c
+++ ares_getopt.c
@@ -66,7 +66,7 @@ int
 ares_getopt(int nargc, char * const nargv[], const char *ostr)
 {
     static char *place = EMSG;                /* option letter processing */
-    char *oli;                                /* option letter list index */
+    const char *oli;                          /* option letter list index */
 
     if (optreset || !*place) {                /* update scanning pointer */
         optreset = 0;
